Update RPC endpoints of Lisk Service - Closes #1574 (#1786)
LiskHQ/lisk-docs: docs/modules/ROOT/pages/api/lisk-service-rpc.adoc

.GET transaction history for an account as a CSV.
[source,json]
----
{
  "jsonrpc": "2.0",
  "id": 1,
  "method": "get.export.transactions",
  "params": {
    "address": "lskdwsyfmcko6mcd357446yatromr9vzgu7eb8y99"
  }
}
----
--
=====

[%collapsible]
.Sample responses
====
TIP: In order to export the account history, follow the steps mentioned in status codes *202* and *200*.

[tabs]
=====
202 ACCEPTED::
+
--
If the request is valid, the server will return with the `address` of the account and the `interval` for which the data export is requested.

[source,json]
----
{
  "data": {
    "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
    "interval": "2021-03-16:2021-12-06",
  },
  "meta": {
    "ready": false
  }
}
----
--
200 OK::
+
--
After a few seconds, re-request the same API with the same credentials to get a URL for the CSV file.

[source,json]
----
{
  "data": {
    "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
    "interval": "2021-03-16:2021-12-06",
    "fileName": "transactions_lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt_2021-03-16_2021-12-06.csv",

>     "fileUrl": "/api/v3/exports/transactions_lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt_2021-03-16_2021-12-06.csv"

Reviewer: ```suggestion
    "fileUrl": "/api/v3/export/download?filename=transactions_lsk9tovmsz2wvewcod7fd99p79wdbswtq9uyrs23t_2021-07-01_2023-05-26.csv"
```

            "publicKey": "a360ead204f95f271bd0a4dc4b36d0cab15fb808cf8205ae10659c4b3bfce367",
            "nextAllocatedTime": 1685095880,
            "status": "active"
        }
    ],
    "meta": {
        "count": 1,
        "offset": 0,
        "total": 1
    }
}
----
--
=====
====

=== Auth
Endpoint related to Auth module.

==== get.auth
Retrieves user details from the Auth module.

[tabs]
=====
Specification::
+
--
[cols="1,1,4,1,4", options="header"]
|===
|Parameter |Datatype |Description |Default |Sample

|`address`
|String
|Address of a blockchain account.
This is a required property.
|none
|lsknhqvv8ou5jpjcn3ezazkjgbjp2kdhwvoyz6hfj

|===

--
RPC Sample Request::
+
--
To request Lisk Service endpoints via WS, set up a *socket.io* connection as suggested in the <<WS, WS-RPC>> intro, and send the following request:

.GET auth details for an address.
[source,json]
----
{
    "jsonrpc": "2.0",
    "id": 1,
    "method": "get.auth",
    "params": {
      "address": "lskdwsyfmcko6mcd357446yatromr9vzgu7eb8y99"
    }
}
----
--

=====

[%collapsible]
.Sample responses
====
[tabs]
=====
200 OK::
+
--
[source,json]
----
{
  "data": {
    "nonce": "1",

>     "numberOfReqSignatures": 1,

Reviewer: ```suggestion
    "numberOfSignatures": 1,
```

|`sort`
|Enum
|Defines the criteria on which the retrieved data is sorted.
Value can be: tokenID:desc, or tokenID:asc.
|tokenID:asc
|tokenID:asc

|`limit`
|Number
|The number of records to be returned for a query.
|10
|10

|`offset`
|Number
|The number of records a query should skip before selecting records.
|0
|0

|===

--
RPC Sample Request::
+
--
To request Lisk Service endpoints via WS, set up a *socket.io* connection as suggested in the <<WS, WS-RPC>> intro, and send the following request:

.GET available token identifiers from the Token module.

[source,json]
----
{
    "jsonrpc": "2.0",
    "id": 1,
    "method": "get.token.available-ids",
    "params": {
    }
}
----
--

=====

[%collapsible]
.Sample responses
====
[tabs]
=====
200 OK::
+
--
[source,json]
----
{
  "data": {
      "tokenIDs": [
          "0000000000000000"
      ]
  },
  "meta": {
      "count": 1,
      "offset": 0,
      "total": 1
  }
}
----
--
=====
====

==== get.token.balances.top
Retrieves top token balances for a token ID.

[tabs]
=====
Specification::
+
--
[cols="1,1,4,1,4", options="header"]
|===
|Parameter |Datatype |Description |Default |Sample

> |`tokenID`

Reviewer: `tokenID` should be required parameter for this endpoint

|none
|lsknhqvv8ou5jpjcn3ezazkjgbjp2kdhwvoyz6hfj

|`limit`
|Number
|The number of records to be returned for a query.
|10
|10

|`offset`
|Number
|The number of records a query should skip before selecting records.
|0
|0

|===

--
RPC Sample Request::
+
--
To request Lisk Service endpoints via WS, set up a *socket.io* connection as suggested in the <<WS, WS-RPC>> intro, and send the following request:

.GET list of stakers.
[source,json]
----
{
    "jsonrpc": "2.0",
    "id": 1,
    "method": "get.pos.stakers",
    "params": {
      "address": "lskdwsyfmcko6mcd357446yatromr9vzgu7eb8y99"
    }
}
----
--

=====

[%collapsible]
.Sample responses
====
[tabs]
=====
200 OK::
+
--
[source,json]
----
{
  "data": {
    "stakers": [
      {
        "address": "lskdwsyfmcko6mcd357446yatromr9vzgu7eb8y99",
        "amount": "10815000000000",
        "name": "liskhq"
      }
    ]
  },
  "meta": {
    "validator": {
      "address": "lskdwsyfmcko6mcd357446yatromr9vzgu7eb8y99",
      "publicKey": "b1d6bc6c7edd0673f5fed0681b73de6eb70539c21278b300f07ade277e1962cd",
      "name": "genesis_84"
    },
    "count": 100,
    "offset": 25,
    "total": 43749
  }
}
----
--
=====
====

==== get.pos.stakes

> Retrieves the list of stakes sent by the specified user by their address, `publicKey`, or validator `name`.

Reviewer: ```suggestion
Retrieves the list of stakes sent by the specified user by their `address`, `publicKey`, or validator `name`.
```

= RPC API of Lisk Service
Muhammad Talha <muhammad.talha@lightcurve.io>
:toc:
:toclevels: 4
:page-toclevels: 4
:url_socket_io: https://socket.io/
:url_connecting_service: integrate-blockchain/connecting-endpoints.adoc

== Introduction to Lisk Service Endpoints

=== API base URL
The base URL for the Lisk Service WS-RPC endpoints are available to query under the `/rpc-v3` namespace.

* *WS-RPC:* `wss://service.lisk.com/rpc-v3/`

=== Structure of endpoint
Each RPC endpoint in Lisk Service follows a specific structure.

For WS, the RPC endpoint name is passed in the `method` property after establishing a WebSocket connection using {url_socket_io}[socket.io].
[#WS]
For more information, see the guide on xref:{url_connecting_service}[].

[source,json]
----
"method": "<requestType>.<moduleName or rootEntity>.<object>.<property>"
----

> The query parameters are passed in the `params` object. For example:

Reviewer 2: ```suggestion
The query parameters or payload are passed in the `params` object. For example:
```

A transaction can be queried based on a nonce as well.
|none
|1

|`limit`
|Number
|The number of records to be returned for a query.
|10
|10

|`offset`
|Number
|The number of records a query should skip before selecting records.
|0
|0

|`sort`
|Enum
|Defines the criteria on which the retrieved data is sorted.
Value can be: height:asc, height:desc, timestamp:asc, or timestamp:desc.
|height:desc
|height:desc

|`order`
|Enum
|The `order` condition applies only after the `sort` condition.
The property is usually used to break ties when the sort condition results in a collision.
Possible values are index:asc or index:desc.
|index:asc
|index:asc

|===
--
RPC Sample Request::
+
--
To request Lisk Service endpoints via WS, set up a *socket.io* connection as suggested in the <<WS, WS-RPC>> intro, and send the following request:

.GET transaction by transactionID.
[source,json]
----
{
  "jsonrpc": "2.0",
  "id": 1,
  "method": "get.transactions",
  "params": {
      "transactionID": "dcb3840032b0bcabc1a0ae5e89124004e537927cfa8e9061d5984eaff91b5243"
  }
}
----

.GET the last 25 transactions for an account.
[source,json]
----
{
  "jsonrpc": "2.0",
  "id": 1,
  "method": "get.transactions",

>   "params": {
>       "address": "lskguo9kqnea2zsfo3a6qppozsxsg92nuuma3p7ad"
>   }

Reviewer: We either need to add `limit:25` to the params or update the description
```suggestion
  "params": {
      "address": "lskguo9kqnea2zsfo3a6qppozsxsg92nuuma3p7ad",
      "limit:" 25
  }
```

----
{
  "transaction":  {
    "module": "token",
    "command": "transferCrossChain",
    "fee": "100000000",
    "nonce": "1",
    "senderPublicKey": "3972849f2ab66376a68671c10a00e8b8b67d880434cc65b04c6ed886dfa91c2c",
    "signatures": [
      "0f0af2be5a18593f76dbd7a5d43e29cb9cce7a056dc28f818cc2d75e671bde9e5cccaf924b2a86415dc49be14c8b6bbf348a8918521b7a028bea1d9637bec905"
    ],
    "params": {
      "tokenID": "0000000000000000",
      "amount": "100000000000",
      "receivingChainID": "00000001",
      "recipientAddress": "lskyvvam5rxyvbvofxbdfcupxetzmqxu22phm4yuo",
      "data": "",
      "messageFee": "10000000",
      "messageFeeTokenID": "0000000000000000"
    },
    "id": "0f77248481c050fcf4f88ef7b967548452869879137364df3b33da09cc419395"
  }
}
----

*OR*

.The encoded payload of a transaction
[source,json]
----
{
  "transaction": "0a040000000212040000000018002080c2d72f2a2044c3cb523c0a069e3f2dcb2d5994b6ba8ff9f73cac9ae746922aac4bc22f95b132310a0800000001000000001080c2d72f1a14632228a3e6a67ac6892de2eb4f60abe2e3bc42a1220a73656e6420746f6b656e3a40964d81e28727e6567b0fcd8a7fcf0a03f401cadbc1c16b9a7f300a52c372022b51a4553865199af34b5f73765f970704fc443d2a6dd510a26748905c306e530b"

> }

Reviewer: This endpoint only accept decoded payload. We can remove this.

        "proofOfPossession": "88bb31b27eae23038e14f9d9d1b628a39f5881b5278c3c6f0249f81ba0deb1f68aa5f8847854d6554051aa810fdf1cdb02df4af7a5647b1aa4afb60ec6d446ee17af24a8a50876ffdaf9bf475038ec5f8ebeda1c1c6a3220293e23b13a9a5d26"
    }
}
----
--
=====

[%collapsible]
.Sample responses
====
[tabs]
=====
200 OK::
+
--
[source,json]
----
{
  "data": {
      "isValid": true
  },
  "meta": {}
}
----
--
=====
====

=== Token
Endpoints related to the Token module.

==== get.token.account.exists
Validates if an entry exists in the Token sub-store for the specified `address`, `name`, or `publicKey`.

[tabs]
=====
Specification::
+
--
[cols="1,1,4,1,4", options="header"]
|===
|Parameter |Datatype |Description |Default |Sample

|`tokenID`
|String
|A unique string value that identifies a token.
This is a required property.
|none
|0000000000000000

|`address`
|String
|Address of a blockchain account.
|none
|lsknhqvv8ou5jpjcn3ezazkjgbjp2kdhwvoyz6hfj

|`publicKey`
|String
|The The public key of a blockchain account.
|none
|b1d6bc6c7edd0673f5fed0681b73de6eb70539c21278b300f07ade277e1962cd

|`name`
|String
|Name of the blockchain account.
|none

> |genesis_84

Reviewer: With tokenID, any one of these properties (name, address or publicKey) is required for this request. We can add comment highlighting this.

[tabs]
=====
200 OK::
+
--
[source,json]
----
{
  "data": {
    "feeEstimatePerByte": {
      "low": 0,
      "medium": 1000,
      "high": 2000
    },
    "feeTokenID": "0000000000000000",
    "minFeePerByte": 1000
  },
  "meta": {
    "lastUpdate": 1616008148,
    "lastBlockHeight": 25,
    "lastBlockID": "01967dba384998026fe028119bd099ecf073c05c045381500a93d1a7c7307e5b"
  }
}
----
--
503 Service Unavailable::
+
--
[source,json]
----
{
  "error": true,
  "message": "Service is not ready yet."
}
----
--
=====
====

=== Proof of Stake (PoS)
Endpoints related to PoS module.

==== get.pos.rewards.claimable
Retrieves currently claimable rewards information from the PoS module for the specified `address`, `publicKey`, or validator `name`.
The API supports pagination.

[tabs]
=====
Specification::
+
--
[cols="1,1,4,1,4", options="header"]
|===
|Parameter |Datatype |Description |Default |Sample

|`address`
|String
|Address of a blockchain account.
|none
|lsknhqvv8ou5jpjcn3ezazkjgbjp2kdhwvoyz6hfj

|`publicKey`
|String
|The public key of a blockchain account.
|none
|b1d6bc6c7edd0673f5fed0681b73de6eb70539c21278b300f07ade277e1962cd

|`name`
|String
|Name of the blockchain account.
|none

> |genesis_84

Reviewer: One of the address, publicKey or name is required here

          "chainName": "Lisk",
          "chainID": "04000000",
          "networkType": "devnet"
      },
      {
          "chainName": "Colecti",
          "chainID": "03000002",
          "networkType": "alphanet"
      },
      {
          "chainName": "Colecti",
          "chainID": "04000002",
          "networkType": "devnet"
      },
      {
          "chainName": "Enevti",
          "chainID": "03000001",
          "networkType": "alphanet"
      },
      {
          "chainName": "Enevti",
          "chainID": "04000001",
          "networkType": "devnet"
      }
  ],
  "meta": {
      "count": 7,
      "offset": 0,
      "total": 7
  }
}
----
--
=====
====

==== get.blockchain.apps.meta
Retrieves metadata for a list of blockchain applications.
The information is used by the wallets.
The API supports pagination.

[tabs]
=====
Specification::
+
--
[cols="1,1,4,1,4", options="header"]
|===
|Parameter |Datatype |Description |Default |Sample

|`chainName`
|String
|Name of the blockchain.
|none
|Lisk

|`chainID`
|String
|A unique string value that identifies a blockchain in a network or various blockchains.
Can be expressed as CSV.
|none
|00000000

|`isDefault`
|Boolean

> |true, false value.

Reviewer: ```suggestion
|Boolean value.
```

    {
      "chainID": "00000000",
      "chainName": "Lisk",
      "tokenID": "Lisk",
      "tokenName": "00000000",
      "networkType": "mainnet",
      "description": "LSK is the utility token of Lisk",
      "denomUnits": [
        {
          "denom": "lsk",
          "decimals": 8,
          "aliases": [
            "LISK"
          ]
        }
      ],
      "symbol": "LSK",
      "displayDenom": "LSK",
      "baseDenom": "beddows",
      "logo": {
        "png": "https://downloads.lisk.com/lisk/images/logo.png",
        "svg": "https://downloads.lisk.com/lisk/images/logo.svg"
      }
    }
  ],
  "meta": {
    "count": 10,
    "offset": 10,
    "total": 400
  }
}
----
--
=====
====

==== get.blockchain.apps.meta.tokens.supported
Retrieves the metadata for the tokens supported by the specified blockchain application.
The information is used by the wallets.
This endpoint internally queries the `/token/summary` for the specified `chainID` and considers the `supportedTokens` information to collate the relevant metadata.
The API supports pagination.

[tabs]
=====
Specification::
+
--
[cols="1,1,4,1,4", options="header"]
|===
|Parameter |Datatype |Description |Default |Sample

> |`chainID`

Reviewer: `chainID` is required parameter here.

}
----
--
=====

[%collapsible]
.Sample responses
====
TIP: In order to export the account history, follow the steps mentioned in status codes *202* and *200*.

[tabs]
=====
202 ACCEPTED::
+
--
If the request is valid, the server will return with the `address` of the account and the `interval` for which the data export is requested.

[source,json]
----
{
  "data": {
    "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
    "interval": "2021-03-16:2021-12-06",
  },
  "meta": {
    "ready": false
  }
}
----
--
200 OK::
+
--
After a few seconds, re-request the same API with the same credentials to get a URL for the CSV file.

[source,json]
----
{
  "data": {
    "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
    "interval": "2021-03-16:2021-12-06",
    "fileName": "transactions_lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt_2021-03-16_2021-12-06.csv",
    "fileUrl": "/api/v3/exports/transactions_lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt_2021-03-16_2021-12-06.csv"
  },
  "meta": {
    "ready": true
  }
}
----

Upon receiving a similar response as mentioned above, combine the `fileUrl` with the HTTP base URL: https://service.lisk.com like the following to download the file.

> https://service.lisk.com/api/v3/exports/transactions_lsk9tovmsz2wvewcod7fd99p79wdbswtq9uyrs23t_2021-07-01_2023-05-26.csv

Reviewer: ```suggestion
https://service.lisk.com/api/v3/export/download?filename=transactions_lsk9tovmsz2wvewcod7fd99p79wdbswtq9uyrs23t_2021-07-01_2023-05-26.csv
```

          },
          "assets": [
              {
                  "module": "random",
                  "data": {
                      "seedReveal": "c6e438bcbfe9e77825cf70e5cafcc2b3"
                  }
              }
          ]
      }
  ],
  "meta": {
      "count": 1,
      "offset": 0,
      "total": 1
  }
}
----
--
=====
====

=== Transactions
API endpoints about transactions.

==== get.transactions
Retrieves network transactions by the criteria defined in the parameters.

[tabs]
=====
Specification::
+
--
[cols="1,1,4,1,4", options="header"]
|===
|Parameter |Datatype |Description |Default |Sample

|`transactionID`
|String
|A unique string value that identifies a transaction.
|none
|dcb3840032b0bcabc1a0ae5e89124004e537927cfa8e9061d5984eaff91b5243

|`moduleCommand`
|String
|Combination of a module and its command name e.g. `transfer` command of the `token` module looks like `token:transfer`.
|none
|token:transfer

|`senderAddress`
|String
|Address of transaction's sender.
|none
|lskguo9kqnea2zsfo3a6qppozsxsg92nuuma3p7ad

|`recipientAddress`
|String
|Address of transaction's receiver.
|none
|lsknhqvv8ou5jpjcn3ezazkjgbjp2kdhwvoyz6hfj

> |`senderPublicKey`

Reviewer: Currently, we don't support querying transactions by `senderPublicKey`. We can remove this.

    }
}
----
--

=====

[%collapsible]
.Sample responses
====
[tabs]
=====
200 OK::
+
--
[source,json]
----
{
  "data": {
    "result": 1,
    "status": "ok",
    "events": [
      {
        "data": {
          "senderAddress": "lskdwsyfmcko6mcd357446yatromr9vzgu7eb8y99",
          "tokenID": "0000000000000000",
          "amount": "100003490",
          "recipientAddress": "lskdwsyfmcko6mcd357446yatromr9vzgu7eb8y99"
        },
        "index": 0,
        "module": "token",
        "name": "transferEvent",
        "topics": [
          "86afcdd640846bf41525481938653ee942be3fac1ecbcff08e98f9aeda3a9583",
          "lskdwsyfmcko6mcd357446yatromr9vzgu7eb8y99",
          "0000000000000000",
          "lskdwsyfmcko6mcd357446yatromr9vzgu7eb8y99"
        ],
        "height": 10
      }
    ]
  },
  "meta": {}
}
----
--
500 Internal Server Error::
+
--
[source,json]
----
{
  "error": true,
  "message": "Unable to reach a network node."
}
----
--
=====
====

==== post.transactions.estimate-fees

> Estimates the fee of a transaction before it is sent to the network node for execution.

Reviewer: ```suggestion
Estimate the fees of a transaction before it is sent to the network node for execution.
```

Author: Does the endpoint allows multiple transactions to be passed in the request?

Reviewer: No, it allows only one transactions but returns different estimated fees for the given transaction. for example, dynamic fee, account initialisation fees, message fee etc.

[tabs]
=====
Specification::
+
--

The endpoint can be invoked by sending a payload to it.

.The decoded payload of a transaction
[source,json]
----
{
  "transaction":  {
    "module": "token",

>     "command": "transfer",

Reviewer: ```suggestion
    "command": "transferCrossChain",
```

[source,json]
----
"method": "<requestType>.<moduleName or rootEntity>.<object>.<property>"
----
The query parameters are passed in the `params` object. For example:

[source,json]
----
{
  "jsonrpc": "2.0",
  "id": 1,
  "method": "get.token.account.exists",
  "params": {
      "tokenID": "0000000000000000"
  }
}
----

=== Response format
All responses are in the JSON format: `application/json`.

Each API request has the following structure:
[source,json]
----
{
  "data": {}, // Contains the requested data
  "meta": {}, // Contains additional metadata, e.g. the values of `limit` and `offset`
}
----

[source,json]
----
{
  "error": true,
  "message": "Unknown input parameter(s): <param_name>", // Contains the error message
}
----

=== API status and readiness
Lisk Service offers `/status` and `/ready` endpoints to check the current deployment and high-level service readiness statuses respectively.
These endpoints must be queried without the API versioning.

* *Example:* https://service.lisk.com/api/status
* *Example:* https://service.lisk.com/api/ready

== Lisk Blockchain-related endpoints

=== Blocks
API endpoints about blocks and block assets.

> ==== get.blocks

Author: Thanks, will make the change as discussed in huddle.

> ==== get.blocks

Reviewer 2: Please note that the HTTP API is already documented on a separate page, so imho it is not necessary to include it here as well

> ==== get.blocks

Author: If you both agree, I can remove the HTTP related content from the page and make it purely RPC related.

> ==== get.blocks

Reviewer: Thank you for the clarification @Tschakki 👍. @TalhaMaliktz I agree to remove all the HTTP related content from this page to make it not confusing.

              "name": "genesis_9",
              "publicKey": null
          },
          "transactionRoot": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
          "assetRoot": "86ac617cde82f4ecfc4597829f4df34634b9dead231e4a22c96152025ef41e4c",
          "stateRoot": "ab4e6af04f93ea0c5348828eb6af2fe8b0c903fe455956c7a48caec6fab306b1",
          "maxHeightPrevoted": 0,
          "maxHeightGenerated": 0,
          "validatorsHash": "8438b6d8c70e18c79a9215f53b6c4677e2f4bab793479a35c03d8f514f3389b3",
          "aggregateCommit": {
              "height": 0,
              "aggregationBits": "",
              "certificateSignature": ""
          },
          "numberOfTransactions": 0,
          "numberOfAssets": 1,
          "numberOfEvents": 1,
          "totalForged": "0",
          "totalBurnt": "0",
          "networkFee": "0",
          "signature": "69ab16efce598facb355cca7441cdb932345e13b50f1f5256efb7f14f1f52fba0d994bbf607803bb0cbf8d1d58169e3b33e8c46fdb5434994b523df42f0b240d",
          "reward": "0",
          "isFinal": true
      }
  ],
  "meta": {
      "count": 1,
      "offset": 0,
      "total": 1
  }
}
----
--
=====
====

> ==== get.blocks.assets

Reviewer: Same as above. Let's add RPC and HTTP endpoint for all endpoints

  "params": {
      "blockID": "485ddcec34ffe77942eb8e4ffb90c323d9eec290aa80ff6fc9806c5679ce0bc2"
  }
}
----

.WS-RPC: GET block assets by height
[source,json]
----
{
  "jsonrpc": "2.0",
  "id": 1,
  "method": "get.blocks.assets",
  "params": {
      "height": "10"
  }
}
----
--

=====

[%collapsible]
.Sample responses
====
[tabs]
=====
200 OK::
+
--
[source,json]
----
{
  "data": [
      {
          "block": {
              "id": "485ddcec34ffe77942eb8e4ffb90c323d9eec290aa80ff6fc9806c5679ce0bc2",
              "height": 10,
              "timestamp": 1683666343
          },
          "assets": [
              {
                  "module": "random",
                  "data": {
                      "seedReveal": "c6e438bcbfe9e77825cf70e5cafcc2b3"
                  }
              }
          ]
      }
  ],
  "meta": {
      "count": 1,
      "offset": 0,
      "total": 1
  }
}
----
--
=====
====

=== Transactions
API endpoints about transactions.

==== get.transactions
Retrieves network transactions by the criteria defined in the parameters.

[tabs]
=====
Specification::
+
--
[cols="1,1,4,1,4", options="header"]
|===
|Parameter |Datatype |Description |Default |Sample

> |`transactionId`

Reviewer: ```suggestion
|`transactionID`
```

|String
|A unique string value that identifies a transaction.
|none
|dcb3840032b0bcabc1a0ae5e89124004e537927cfa8e9061d5984eaff91b5243

|`moduleCommand`
|String

> |Command of a module e.g. `transfer` command of the token module.

Reviewer: ```suggestion
|Combination of transaction moduleName:commandName.
```

[source,json]
----
{
  "jsonrpc": "2.0",
  "id": 1,
  "method": "get.transactions",
  "params": {
      "transactionID": "dcb3840032b0bcabc1a0ae5e89124004e537927cfa8e9061d5984eaff91b5243"
  }
}
----

.WS-RPC: GET the last 25 transactions for an account.
[source,json]
----
{
  "jsonrpc": "2.0",
  "id": 1,
  "method": "get.transactions",
  "params": {
      "address": "lskguo9kqnea2zsfo3a6qppozsxsg92nuuma3p7ad"
  }
}
----
--

=====

[%collapsible]
.Sample responses
====
[tabs]
=====
200 OK::
+
--
[source,json]
----
{
  "data": [
      {
          "id": "dcb3840032b0bcabc1a0ae5e89124004e537927cfa8e9061d5984eaff91b5243",
          "moduleCommand": "token:transfer",
          "nonce": "4",
          "fee": "5174000",
          "minFee": "173000",
          "size": 174,
          "sender": {
              "address": "lskguo9kqnea2zsfo3a6qppozsxsg92nuuma3p7ad",
              "publicKey": "3972849f2ab66376a68671c10a00e8b8b67d880434cc65b04c6ed886dfa91c2c",
              "name": "genesis_0"
          },
          "params": {
              "tokenID": "0000000000000000",
              "amount": "12000000000",
              "recipientAddress": "lsknhqvv8ou5jpjcn3ezazkjgbjp2kdhwvoyz6hfj",

>               "data": "welcome "

Reviewer: ```suggestion
              "data": "welcome"
```

          },
          "block": {
              "id": "5d81effdba82a15977935609e4626091ee904f289e5d7074d67a5b26750064b9",
              "height": 880,
              "timestamp": 1684325210,
              "isFinal": false
          },
          "meta": {
              "recipient": {
                  "address": "lsknhqvv8ou5jpjcn3ezazkjgbjp2kdhwvoyz6hfj",
                  "publicKey": null,
                  "name": null
              }
          },
          "executionStatus": "success",
          "index": 0
      }
  ],
  "meta": {
      "count": 1,
      "offset": 0,
      "total": 1
  }
}
----
--
=====
====

==== post.transactions.dryrun
Sends decoded or encoded transaction(s) to the node for dry-running purposes.

[tabs]
=====
Specification::
+
--
No parameters are required.
The endpoint can be invoked by sending a payload to it.

.The decoded payload of a transaction.
[source,json]
----
{
  "skipDecode": false,

>   "skipVerify": false,

Reviewer: May be we can add a note to clarify these 2 properties:
skipVerify: A boolean indicator to skip the transaction verification
skipDecode: A boolean indicator to skip the auto-decoding of the event data

--header 'Content-Type: application/json' \
--data '{
  "transaction":  {
    "module": "token",
    "command": "transfer",
    "fee": "100000000",
    "nonce": "1",
    "senderPublicKey": "3972849f2ab66376a68671c10a00e8b8b67d880434cc65b04c6ed886dfa91c2c",
    "signatures": [
      "0f0af2be5a18593f76dbd7a5d43e29cb9cce7a056dc28f818cc2d75e671bde9e5cccaf924b2a86415dc49be14c8b6bbf348a8918521b7a028bea1d9637bec905"
    ],
    "params": {
      "tokenID": "0000000000000000",
      "amount": "100000000000",
      "receivingChainID": "00000001",
      "recipientAddress": "lskyvvam5rxyvbvofxbdfcupxetzmqxu22phm4yuo",
      "data": "",
      "messageFee": "10000000",
      "messageFeeTokenID": "0000000000000000"
    },
    "id": "0f77248481c050fcf4f88ef7b967548452869879137364df3b33da09cc419395"
  }
}'
----
--
WS-RPC::
+
--
To request Lisk Service endpoints via WS, set up a *socket.io* connection as suggested in the <<WS, WS-RPC>> intro, and send the following request:

.WS-RPC: POST payload of a transaction to estimate fee.
[source,json]
----
{
  "jsonrpc": "2.0",
  "id": 1,
  "method": "post.transactions.estimate-fees",
  "params": {
      "transaction": {
          "module": "token",

>           "command": "transfer",

Reviewer: ```suggestion
          "command": "transferCrossChain",
```

          "fee": "100000000",
          "nonce": "1",
          "senderPublicKey": "3972849f2ab66376a68671c10a00e8b8b67d880434cc65b04c6ed886dfa91c2c",
          "signatures": [
              "0f0af2be5a18593f76dbd7a5d43e29cb9cce7a056dc28f818cc2d75e671bde9e5cccaf924b2a86415dc49be14c8b6bbf348a8918521b7a028bea1d9637bec905"
          ],
          "params": {
              "tokenID": "0000000000000000",
              "amount": "100000000000",
              "receivingChainID": "00000001",
              "recipientAddress": "lskyvvam5rxyvbvofxbdfcupxetzmqxu22phm4yuo",
              "data": "",
              "messageFee": "10000000",
              "messageFeeTokenID": "0000000000000000"
          },
          "id": "0f77248481c050fcf4f88ef7b967548452869879137364df3b33da09cc419395"
      }
  }
}
----
--

=====

[%collapsible]
.Sample responses
====
[tabs]
=====
200 OK::
+
--
[source,json]
----
{
  "data": {

>       "transactionFeeEstimates": {
>           "minFee": "166000",
>           "accountInitializationFee": {
>               "tokenID": "0000000000000000",
>               "amount": "5000000"
>           }
>       },

Reviewer: Let's update the example response as request payload is for token:transferCrossChain transaction:
```suggestion
      "transactionFeeEstimates": {
          "minFee": "166000",
          "accountInitializationFee": {
              "tokenID": "0000000000000000",
              "amount": "5000000"
          },
          "messageFee": {
              "tokenID": "0000000000000000",
              "amount": "343000"
          }
      },
```

|10
|10

|`offset`
|Number
|The number of records a query should skip before selecting records.
|0
|0

|===

--
cURL::
+
--
.HTTP: GET transaction statistics for the past 7 days.
[source,json]
----
curl --location 'https://service.lisk.com/api/v3/transactions/statistics?interval=day&limit=7'
----

.HTTP: GET transaction statistics for the past 12 months.
[source,json]
----
curl --location 'https://service.lisk.com/api/v3/transactions​/statistics​?interval=month&limit=12'
----
--
WS-RPC::
+
--
To request Lisk Service endpoints via WS, set up a *socket.io* connection as suggested in the <<WS, WS-RPC>> intro, and send the following request:

.WS-RPC: GET transaction statistics for the past 7 days.
[source,json]
----
{
  "jsonrpc": "2.0",
  "id": 1,
  "method": "get.transactions.statistics",
  "params": {
      "interval": "day",
      "limit": 7
  }
}
----
--

=====

[%collapsible]
.Sample responses
====
[tabs]
=====
200 OK::
+
--
[source,json]
----
{
  "data": {
      "distributionByType": {
          "pos:changeCommission": 1,
          "pos:registerValidator": 2,
          "pos:stake": 8,
          "token:transfer": 9
      },
      "distributionByAmount": {

>           "0200000000000000": {

Reviewer: Lets try to use `mainnet` tokenID everywhere
```suggestion
          "0000000000000000": {
```

                  "transactionCount": 0,
                  "volume": 0,
                  "timestamp": 1683936000
              },
              {
                  "date": "2023-05-12",
                  "transactionCount": 0,
                  "volume": 0,
                  "timestamp": 1683849600
              },
              {
                  "date": "2023-05-11",
                  "transactionCount": 0,
                  "volume": 0,
                  "timestamp": 1683763200
              }
          ]
      }
  },
  "meta": {
      "limit": 7,
      "offset": 0,
      "total": 365,
      "duration": {
          "format": "YYYY-MM-DD",
          "from": "2023-05-11",
          "to": "2023-05-17"
      }
  }
}
----
--
503 Service Unavailable::
+
--
[source,json]
----
{
  "error": true,
  "message": "Service is not ready yet."
}
----
--
=====
====

=== Events
Endpoint related to blockchain events.

==== get.events
Retrieves blockchain events based on the criteria defined by parameters.
The API supports pagination.

[tabs]
=====
Specification::
+
--
[cols="1,1,4,1,4", options="header"]
|===
|Parameter |Datatype |Description |Default |Sample

> |`transactionId`

Reviewer: ```suggestion
|`transactionID`
```

|String
|A unique string value that identifies a transaction.
|none
|dcb3840032b0bcabc1a0ae5e89124004e537927cfa8e9061d5984eaff91b5243

|`senderAddress`
|String
|Address of transaction's sender.
|none
|lskguo9kqnea2zsfo3a6qppozsxsg92nuuma3p7ad

|`topic`
|String
|Highlights the entities an event relates to.
A topic aids in searching for a relevant event.

> Search can be done by passing an `address` or `transactionID` to the topic 

Reviewer: ```suggestion
Search can be done by passing an `address` or `transactionID` to the topic.
```

    "optionalKeys": [
      "478842a844914f18a1c620a6494bf9931d0a862e96212bf5fc6f3ca18658febe"
    ]
  },
  "meta": {
    "address": "lskdwsyfmcko6mcd357446yatromr9vzgu7eb8y99",
    "publicKey": "b1d6bc6c7edd0673f5fed0681b73de6eb70539c21278b300f07ade277e1962cd",
    "name": "genesis_84"
  }
}
----
--
=====
====

=== Validator
Endpoint related to validators.

==== get.validator
Retrieves user details from the Validator module.

[tabs]
=====
Specification::
+
--
[cols="1,1,4,1,4", options="header"]
|===
|Parameter |Datatype |Description |Default |Sample

|`address`
|String
|Address of a blockchain account.
This is a required property.
|none
|lsknhqvv8ou5jpjcn3ezazkjgbjp2kdhwvoyz6hfj

|===

--
cURL::
+
--
.HTTP: GET user details for an address.
[source,json]
----
curl --location 'https://service.lisk.com/api/v3/validator?address=lskc94hx5zgu3vbydwj96v64r5pdfj4q3ytfxyw23'
----
--
WS-RPC::
+
--
To request Lisk Service endpoints via WS, set up a *socket.io* connection as suggested in the <<WS, WS-RPC>> intro, and send the following request:

.WS-RPC: GET user details for an address.
[source,json]
----
{
    "jsonrpc": "2.0",
    "id": 1,
    "method": "get.validator",
    "params": {

>       "address": "lskc94hx5zgu3vbydwj96v64r5pdfj4q3ytfxyw23"

Reviewer: ```suggestion
      "address": "lskdwsyfmcko6mcd357446yatromr9vzgu7eb8y99"
```

Author: Why change this? 🤔

                      },
                      "optionalKeys": {
                          "type": "array",
                          "items": {
                              "dataType": "bytes",
                              "minLength": 32,
                              "maxLength": 32
                          },
                          "fieldNumber": 5
                      }
                  }
              }
          }
      ]
  },
  "meta": {}
}
----
--
=====
====

=== Interoperability
Endpoints related to the Interoperability module.

==== get.blockchain.apps
Retrieves blockchain applications in the current network.
The API supports pagination.

[tabs]
=====
Specification::
+
--
[cols="1,1,4,1,4", options="header"]
|===
|Parameter |Datatype |Description |Default |Sample

|`chainID`
|String
|A unique string value that identifies a blockchain in a network or various blockchains.
Can be expressed as CSV.
|none
|00000000

|`name`
|String
|Name of the blockchain.
|none
|Lisk

|`status`
|String
|Status of a blockchain.
Possible value can be: `registered`, `active`, `terminated`, or `unregistered`.
Can be expressed as CSV.
|none
|active

|`search`
|String

> |Case-insensitive keyword search by `name`.

Reviewer: ```suggestion
|Case-insensitive keyword search by `chainName`.
```

The API supports both partial and full-text search.
|none
|lsknhqvv8ou5jpjcn3ezazkjgbjp2kdhwvoyz6hfj

|`limit`
|Number
|The number of records to be returned for a query.
|10
|10

|`offset`
|Number
|The number of records a query should skip before selecting records.
|0
|0

|===

--
cURL::
+
--
.HTTP: GET blockchain applications in the current network.
[source,json]
----
curl --location 'https://service.lisk.com/api/v3/blockchain/apps'
----
--
WS-RPC::
+
--
To request Lisk Service endpoints via WS, set up a *socket.io* connection as suggested in the <<WS, WS-RPC>> intro, and send the following request:

.WS-RPC: GET blockchain applications in the current network.
[source,json]
----
{
    "jsonrpc": "2.0",
    "id": 1,
    "method": "get.blockchain.apps",
    "params": {
    }
}
----
--

=====

[%collapsible]
.Sample responses
====
[tabs]
=====
200 OK::
+
--
[source,json]
----
{
  "data": [
    {

>       "name": "Lisk",

Reviewer: ```suggestion
      "chainName": "Lisk",
```

  }
}
----

=== Response format
All responses are in the JSON format: `application/json`.

Each API request has the following structure:
[source,json]
----
{
  "data": {}, // Contains the requested data
  "meta": {}, // Contains additional metadata, e.g. the values of `limit` and `offset`
}
----

[source,json]
----
{
  "error": true,
  "message": "Unknown input parameter(s): <param_name>", // Contains the error message
}
----

=== API status and readiness
Lisk Service offers `/status` and `/ready` endpoints to check the current deployment and high-level service readiness statuses respectively.
These endpoints must be queried without the API versioning.

* *Example:* https://service.lisk.com/api/status
* *Example:* https://service.lisk.com/api/ready

== Lisk Blockchain-related endpoints

=== Blocks
API endpoints about blocks and block assets.

==== get.blocks
Retrieves blocks from the blockchain based on properties such as `blockID`, `height`, `account`, etc.
The API supports pagination.

[tabs]
=====
Specification::
+
--
[cols="1,1,2,1,2", options="header"]
|===
|Parameter |Datatype |Description |Default |Sample

|`blockID`
|String
|Block ID of a block in the blockchain.
|none

> |01967dba384998026fe028119bd099ecf073c05c045381500a93d1a7c7307e5b

Author: Thanks, the issue you mention is valid and to mitigate this, I have updated the width of columns to make more space for the content. Due to that, text-wrapping has improved. For strings like `blockID`, `publicKey`, `address`, etc. Antora manages okay and IMHO we shouldn't break these strings for better maintainability and easy copying of text if in case required.

In the past, I have broken strings such as transaction hash in the RPC API for Lisk nodes page coz that's where Antora fails to render an extremely long string. For stylistic preference and synchronicity of content throughout the page, I am avoiding the use of monospace.

|`height`
|String
|Height of a block in the blockchain.
Query by height e.g. 10 or a range of height.
Can be expressed as an interval i.e. `1:20` or `1:` or `:20`.
Specified values are inclusive.
| none
| 10

|`timestamp`
|String
|Query by timestamp or by the range of timestamps.
Can be expressed as an interval i.e. `1000000:2000000`, `1000000:` or `:2000000`.
Specified values are inclusive.
|none
|1000000:2000000

> |`generatorAddress`

Author: Please consider the following URL for new API content:
https://github.com/LiskHQ/lisk-service/blob/release/0.7.0/docs/api/version3.md#request-parameters

              "name": "genesis_0"
          },
          "params": {
              "tokenID": "0400000000000000",
              "amount": "12000000000",
              "recipientAddress": "lsknhqvv8ou5jpjcn3ezazkjgbjp2kdhwvoyz6hfj",
              "data": "welcome "
          },
          "block": {
              "id": "5d81effdba82a15977935609e4626091ee904f289e5d7074d67a5b26750064b9",
              "height": 880,
              "timestamp": 1684325210,
              "isFinal": false
          },
          "meta": {
              "recipient": {
                  "address": "lsknhqvv8ou5jpjcn3ezazkjgbjp2kdhwvoyz6hfj",
                  "publicKey": null,
                  "name": null
              }
          },
          "executionStatus": "success",
          "index": 0
      }
  ],
  "meta": {
      "count": 1,
      "offset": 0,
      "total": 1
  }
}
----
--
400 Bad Request::
+
--
[source,json]
----
{
  "error": true,
  "message": "Unknown input parameter(s): <param_name>"
}
----
--
=====
====

==== post.transactions.dryrun
Sends decoded or encoded transaction(s) to the node for dry-running purposes.

[tabs]
=====
Specification::
+
--
No parameters are required.

> The endpoint can be invoked by sending a payload to it.

Author: AFAIK No! Payload is passed into the body of a `POST` request, whereas params are passed in the URL of a `GET` request. For more information see [Parameters and Payload of an Operation](https://learn.openapis.org/specification/parameters.html#:~:text=Parameters%20are%20typically%20used%20to,provides%20content%20for%20that%20resource.).

> Parameters are typically used to identify a resource, whereas the message payload provides content for that resource.

> The endpoint can be invoked by sending a payload to it.

Reviewer 2: I see, but then the naming is a little confusing, because we specify the payload as params, e.g. here: https://github.com/LiskHQ/lisk-docs/pull/1786/files#diff-4ae369dcd626f789e3bf5b28d232374fec7d3da314be9bf5a2c987cec481ef97R989
